### PR TITLE
Game over screen

### PR DESCRIPTION
**Pull request by chern042**
Created game over screen, as well as implemented a manager for the different UI elements in order to remove/add them from view. This includes the UI elements for the health bar, mana bar, and inventory. I also added a very simple pause menu, and made it so you could restart the scene after death.

### CODE REVIEW

**Comment by RossBaumgartner** [comment]
✔ Pause and game over screens basic functionality working as expected

Bugs Found:

❌Spamming Esc and WASD will lock up the game, then to un-lock up the game you have to keep pausing to let the enemy make its turn (?) Not a critical bug but something to look into

❌Health Bar is not properly working

Other:

❗ There are 3 conflicting files
        - The player controller one is probably the only actual conflict

🚧The branch looks old, maybe update it with a merge in from master? idk if thats worth doing tho

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Controllers/PlayerController.cs:46]: This change needs to be undone, it makes the player invincible.

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Controllers/PlayerController.cs:35]: This block should not be indented.

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Controllers/PlayerController.cs:28]: Pls remove commented out code.

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Controllers/PlayerController.cs:83]: Pls remove extra whitespace

File: Assets/Scripts/Controllers/CharController.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Controllers/CharController.cs:2]: Unused import, better not to have a dependency here if we don't need it.

File: Assets/Scripts/HealthBar.cs
Comment (AndreiBaetica) [inline Assets/Scripts/HealthBar.cs:30]: Pls remove extra whitespace

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Controllers/PlayerController.cs:1]: Unused imports here need removal

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Controllers/PlayerController.cs:129]: Couple of things about this method:

- Why set the player turn to false on death?
- The code block 132-138 is an exact duplicate of what we have in the parent method. Instead of doing this call base.Die()

File: Assets/Scripts/Managers/OverlayManager.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Managers/OverlayManager.cs:79]: Functions here must be capitalized, unlike in Java

File: Assets/player.cs
Comment (AndreiBaetica) [inline Assets/player.cs:1]: Is this class being used?

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Controllers/PlayerController.cs:61]: Why handle death call here as well? We already handle this logic here:

https://github.com/AndreiBaetica/Dungeoneer/blob/master/Assets/Scripts/Controllers/CharController.cs#L217-L221

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Controllers/PlayerController.cs:32]: Since we don't have a defined way of tracking progress yet, can we get rid of this for now?

For when we do end up tracking such values, it should happen somewhere else, not in PlayerController

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Controllers/PlayerController.cs:128]: Pls remove extra whitespace

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (AndreiBaetica) [inline Assets/Scripts/Controllers/PlayerController.cs:144]: Pls remove extra whitespace

File: Assets/prefabs/GameManager.prefab.meta
Comment (AndreiBaetica) [inline Assets/prefabs/GameManager.prefab.meta:1]: Not sure this should be a prefab, at least not yet. Eventually for when we do it programatically I reckon it'll be better to bundle all managers into one prefab and use that in the scene.

Thoughts?

File: Assets/prefabs/GameManager.prefab
Comment (AndreiBaetica) [inline Assets/prefabs/GameManager.prefab:1]: Not sure this should be a prefab, at least not yet. Eventually for when we do it programatically I reckon it'll be better to bundle all managers into one prefab and use that in the scene.

Thoughts?

File: Assets/Scripts/HealthBar.cs
Comment (AndreiBaetica) [inline Assets/Scripts/HealthBar.cs:12]: Is this here to track player health? If so we already do it here:

https://github.com/AndreiBaetica/Dungeoneer/blob/master/Assets/Scripts/Controllers/CharController.cs#L12

File: Assets/Scripts/ButtonManager.prefab
Comment (AndreiBaetica) [inline Assets/Scripts/ButtonManager.prefab:1]: Prefabs should not be in the scripts folder. Also, this comment:
https://github.com/AndreiBaetica/Dungeoneer/pull/133/files?file-filters%5B%5D=.prefab#r667295445

As an aside, I also see another prefab here called DamageIndicatorParent. It should be moved to the prefabs folder.

**Comment by AndreiBaetica** [comment]
> ✔ Pause and game over screens basic functionality working as expected
> 
> Bugs Found:
> 
> ❌Spamming Esc and WASD will lock up the game, then to un-lock up the game you have to keep pausing to let the enemy make its turn (?) Not a critical bug but something to look into
> 
> ❌Health Bar is not properly working
> 
> Other:
> 
> ❗ There are 3 conflicting files
> - The player controller one is probably the only actual conflict
> 
> 🚧The branch looks old, maybe update it with a merge in from master? idk if thats worth doing tho

@RossBaumgartner I was unable to replicate the spamming esc bug, do you know of a specific button sequence that causes this?

Branch is definitely old though, master should be merged here before we merge this into master. It's always worth doing, else we risk letting outdated code/bugs slip by

**Review by AndreiBaetica — CHANGES_REQUESTED** [review]
(see previous comments)
The main thing is that the branch is out of date, so that needs to be updated first.
Other than that, a few bugs and potetially un-necessary code made it in.
Lastly, some minor formatting is required. Duplicate code, whitespace, and moving some files around.

**Comment by RossBaumgartner** [comment]
> @RossBaumgartner I was unable to replicate the spamming esc bug, do you know of a specific button sequence that causes this?
> 
> Branch is definitely old though, master should be merged here before we merge this into master. It's always worth doing, else we risk letting outdated code/bugs slip by

@AndreiBaetica no specific button combo to replicate, I was just spamming the wasd and esc keys, but it was pretty consistantly reproducable on my end

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (chern042) [reply Assets/Scripts/Controllers/PlayerController.cs:126]: It's not a duplicate. Player dying triggers the game over screen, but not the bandit dying. The override contains OverlayManager.Endgame()

File: Assets/Scripts/Controllers/PlayerController.cs
Comment (chern042) [reply Assets/Scripts/Controllers/PlayerController.cs:29]: No

**Review by SebastienGorman — APPROVED** [review]
Tested and looks good 👍. Please wait for Andrei's review at least before merging @chern042.

**Pull request closed without merging**
